### ISSUE TITLE
Nvidia DisplayPort output produces no audio

### ISSUE THREAD

**Issue by Philipp91** [jackaudio/jack2#746]
## Describe the bug

My Nvidia DisplayPort audio output does not work with Jack, but it works with PulseAudio and my onboard audio works with both. I'm seeing messages like `XRUN callback (19 skipped)` continuously.

## Environment

* **JACK Version:** jackdmp version 1.9.12 tmpdir /dev/shm protocol 8
* **Operating System:** Linux Mint 20.1 Cinnamon, kernel 5.4.0-70-generic
* **Nvidia driver:** version 460.39-0ubuntu0.20.04.1
* **Installation:** jack2 installed from package manager (Mint with KXStudio repositories)

I have an Nvidia GeForce GTX 745 card, a monitor connected over DisplayPort and headphones connected to it. I also have a regular onboard sound card.

```
$ aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: PCH [HDA Intel PCH], device 0: ALC891 Analog [ALC891 Analog]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: NVidia [HDA NVidia], device 3: HDMI 0 [HDMI 0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: NVidia [HDA NVidia], device 7: HDMI 1 [HDMI 1]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: NVidia [HDA NVidia], device 8: HDMI 2 [HDMI 2]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: NVidia [HDA NVidia], device 9: HDMI 3 [HDMI 3]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: NVidia [HDA NVidia], device 10: HDMI 4 [HDMI 4]
  Subdevices: 1/1
  Subdevice #0: subdevice #0

$ aplay -L
default
    Playback/recording through the PulseAudio sound server
surround21
    2.1 Surround output to Front and Subwoofer speakers
surround40
    4.0 Surround output to Front and Rear speakers
surround41
    4.1 Surround output to Front, Rear and Subwoofer speakers
surround50
    5.0 Surround output to Front, Center and Rear speakers
surround51
    5.1 Surround output to Front, Center, Rear and Subwoofer speakers
surround71
    7.1 Surround output to Front, Center, Side, Rear and Woofer speakers
null
    Discard all samples (playback) or generate zero samples (capture)
samplerate
    Rate Converter Plugin Using Samplerate Library
speexrate
    Rate Converter Plugin Using Speex Resampler
jack
    JACK Audio Connection Kit
oss
    Open Sound System
pulse
    PulseAudio Sound Server
upmix
    Plugin for channel upmix (4,6,8)
vdownmix
    Plugin for channel downmix (stereo) with a simple spacialization
sysdefault:CARD=PCH
    HDA Intel PCH, ALC891 Analog
    Default Audio Device
front:CARD=PCH,DEV=0
    HDA Intel PCH, ALC891 Analog
    Front speakers
surround21:CARD=PCH,DEV=0
    HDA Intel PCH, ALC891 Analog
    2.1 Surround output to Front and Subwoofer speakers
surround40:CARD=PCH,DEV=0
    HDA Intel PCH, ALC891 Analog
    4.0 Surround output to Front and Rear speakers
surround41:CARD=PCH,DEV=0
    HDA Intel PCH, ALC891 Analog
    4.1 Surround output to Front, Rear and Subwoofer speakers
surround50:CARD=PCH,DEV=0
    HDA Intel PCH, ALC891 Analog
    5.0 Surround output to Front, Center and Rear speakers
surround51:CARD=PCH,DEV=0
    HDA Intel PCH, ALC891 Analog
    5.1 Surround output to Front, Center, Rear and Subwoofer speakers
surround71:CARD=PCH,DEV=0
    HDA Intel PCH, ALC891 Analog
    7.1 Surround output to Front, Center, Side, Rear and Woofer speakers
dmix:CARD=PCH,DEV=0
    HDA Intel PCH, ALC891 Analog
    Direct sample mixing device
dsnoop:CARD=PCH,DEV=0
    HDA Intel PCH, ALC891 Analog
    Direct sample snooping device
hw:CARD=PCH,DEV=0
    HDA Intel PCH, ALC891 Analog
    Direct hardware device without any conversions
plughw:CARD=PCH,DEV=0
    HDA Intel PCH, ALC891 Analog
    Hardware device with all software conversions
usbstream:CARD=PCH
    HDA Intel PCH
    USB Stream Output
hdmi:CARD=NVidia,DEV=0
    HDA NVidia, HDMI 0
    HDMI Audio Output
hdmi:CARD=NVidia,DEV=1
    HDA NVidia, HDMI 1
    HDMI Audio Output
hdmi:CARD=NVidia,DEV=2
    HDA NVidia, HDMI 2
    HDMI Audio Output
hdmi:CARD=NVidia,DEV=3
    HDA NVidia, HDMI 3
    HDMI Audio Output
hdmi:CARD=NVidia,DEV=4
    HDA NVidia, HDMI 4
    HDMI Audio Output
dmix:CARD=NVidia,DEV=3
    HDA NVidia, HDMI 0
    Direct sample mixing device
dmix:CARD=NVidia,DEV=7
    HDA NVidia, HDMI 1
    Direct sample mixing device
dmix:CARD=NVidia,DEV=8
    HDA NVidia, HDMI 2
    Direct sample mixing device
dmix:CARD=NVidia,DEV=9
    HDA NVidia, HDMI 3
    Direct sample mixing device
dmix:CARD=NVidia,DEV=10
    HDA NVidia, HDMI 4
    Direct sample mixing device
dsnoop:CARD=NVidia,DEV=3
    HDA NVidia, HDMI 0
    Direct sample snooping device
dsnoop:CARD=NVidia,DEV=7
    HDA NVidia, HDMI 1
    Direct sample snooping device
dsnoop:CARD=NVidia,DEV=8
    HDA NVidia, HDMI 2
    Direct sample snooping device
dsnoop:CARD=NVidia,DEV=9
    HDA NVidia, HDMI 3
    Direct sample snooping device
dsnoop:CARD=NVidia,DEV=10
    HDA NVidia, HDMI 4
    Direct sample snooping device
hw:CARD=NVidia,DEV=3
    HDA NVidia, HDMI 0
    Direct hardware device without any conversions
hw:CARD=NVidia,DEV=7
    HDA NVidia, HDMI 1
    Direct hardware device without any conversions
hw:CARD=NVidia,DEV=8
    HDA NVidia, HDMI 2
    Direct hardware device without any conversions
hw:CARD=NVidia,DEV=9
    HDA NVidia, HDMI 3
    Direct hardware device without any conversions
hw:CARD=NVidia,DEV=10
    HDA NVidia, HDMI 4
    Direct hardware device without any conversions
plughw:CARD=NVidia,DEV=3
    HDA NVidia, HDMI 0
    Hardware device with all software conversions
plughw:CARD=NVidia,DEV=7
    HDA NVidia, HDMI 1
    Hardware device with all software conversions
plughw:CARD=NVidia,DEV=8
    HDA NVidia, HDMI 2
    Hardware device with all software conversions
plughw:CARD=NVidia,DEV=9
    HDA NVidia, HDMI 3
    Hardware device with all software conversions
plughw:CARD=NVidia,DEV=10
    HDA NVidia, HDMI 4
    Hardware device with all software conversions
usbstream:CARD=NVidia
    HDA NVidia
    USB Stream Output
sysdefault:CARD=STUDIOLOGICMD
    Default Audio Device
usbstream:CARD=STUDIOLOGICMD
    STUDIOLOGIC-MD
    USB Stream Output
```

When I'm not using Jack but just the default PulseAudio installation, my headphones work both with the onboard soundcard and with the DisplayPort output. The latter is called "HDMI / DisplayPort 2 GM107 High Definition Audio Controller [GeForce 940MX]" -> "Digital Stereo (HDMI 2) Output" in the Mint settings dialog. From that "HDMI 2" and my `aplay -l` output, I figure that it's using device number 8. (Small caveat: The onboard soundcard works better with PulseAudio in that the DisplayPort output has some audio glitches/crackles whenever there's a lot going on in the video stream. Nevertheless it works mostly fine, and I'm interested in using the DisplayPort output always because it means less plugging for me and it achieved lower latencies under Windows at least).

Qjackctl lists: hw:PCH, hw:PCH,0, hw:NVidia, hw:NVidia,3, hw:NVidia,7, hw:NVidia,8, hw:NVidia,9, hw:NVidia,10.

## Steps To Reproduce

I've tried all those devices through Qjackctl as well as just on the command line:

```
Turn off PulseAudio on the HDMI output:
$ pactl suspend-sink alsa_output.pci-0000_01_00.1.hdmi-stereo-extra1 1
$ jackd -d alsa -d hw:NVidia -P hw:NVidia,8

On another terminal:
$ jack_simple_client
```

The onboard card (hw:PCH) works (I hear a test sound), but none of the hw:NVidia devices work with Jack. The jackd output is:

```
$ jackd -d alsa -d hw:NVidia -P hw:NVidia,8
jackdmp 1.9.12
Copyright 2001-2005 Paul Davis and others.
Copyright 2004-2016 Grame.
Copyright 2016-2017 Filipe Coelho.
jackdmp comes with ABSOLUTELY NO WARRANTY
This is free software, and you are welcome to redistribute it
under certain conditions; see the file COPYING for details
no message buffer overruns
no message buffer overruns
no message buffer overruns
JACK server starting in realtime mode with priority 10
self-connect-mode is "Don't restrict self connect requests"
audio_reservation_init
Acquire audio card Audio1
creating alsa driver ... hw:NVidia,8|-|1024|2|48000|0|0|nomon|swmeter|-|32bit
configuring for 48000Hz, period = 1024 frames (21.3 ms), buffer = 2 periods
ALSA: final selected sample format for playback: 32bit integer little-endian
ALSA: use 2 periods for playback
```

With Qjackctl, I get additional log output: Every other second, it prints an xrun callback error:

```
19:51:34.316 XRUN callback (17 skipped).
19:51:36.325 XRUN callback (78 skipped).
19:51:38.332 XRUN callback (76 skipped).
19:51:40.343 XRUN callback (79 skipped).
19:51:42.352 XRUN callback (74 skipped).
19:51:44.361 XRUN callback (80 skipped).
19:51:46.370 XRUN callback (75 skipped).
...
```

### Expected vs. actual behavior

I expect to hear audio output, but the headphones connected to the DisplayPort monitor remain silent when using Jack.

**Comment by falkTX** [comment]
likely just a matter of increasing the buffer size.
1024 might be too little for HDMI.

**Comment by Philipp91** [comment]
Right, I had tried that too, I had previously tried lower numbers (on Windows with ASIO4All it's stable with 64 or 128). 1024 doesn't work (yet) and 2048 gives: "ERROR: ALSA: cannot set period size to 2048 frames for playback"

**Comment by Philipp91** [comment]
I've switched to HDMI, which had the same issue initially. But HDMI doesn't have crackles when I watch YouTube videos, so I'm sticking with that. I found a solution for Jack with HDMI now, not sure if it would also have worked on DisplayPort.

The solution was to add `options snd-hda-intel bdl_pos_adj=48,48,48` (or `64,64,64` also works fine) in `/etc/modprobe.d/alsa-base.conf`. It may be worth playing with these values -- in my case, `24,24,24` or lower _and_ `72,72,72` or higher all don't work, it really has to be that sweet middle ground. Then a buffer size of 256 (or even 128) works fine for me with Jack with its realtime setting. I have a lowlatency kernel.
